README.md
OOP Game
======================================

[![Conventional Commits][conventional-commits-image]][conventional-commits-url]

The scope of the project is about to highlight Oriented Object Programming concepts and S.O.L.I.D principles.

## About the project
#### Project stack  
* PHP 7.2
* nginx alpine
* Composer
* Phpunit
* Bootstrap 4

#### Design patterns
* Singleton
* Front Controller
* MVC
* Factory Method
* Fluent Interface
* Dependency Injection
* Strategy
* Delegation

#### Story and Project requirements  
**The story**  
Once upon a time there was a great hero, called **Orderus**, with some strengths and weaknesses, as all heroes have.
After battling all kinds of monsters for more than a hundred years, **Orderus** now has the following stats:
* Health: 70 - 100
* Strength: 70 - 80
* Defence: 45 – 55
* Speed: 40 – 50
* Luck: 10% - 30% (0% means no luck, 100% lucky all the time)  

Also, he possesses 2 skills:
* Rapid strike: Strike twice while it’s his turn to attack; there’s a 10% chance he’ll use this skill every time he attacks
* Magic shield: Takes only half of the usual damage when an enemy attacks; there’s a 20% change he’ll use this skill every time he defends.

**Gameplay**  
As **Orderus** walks the ever-green forests of **Emagia**, he encounters some wild beasts, with the following properties:
* Health: 60 - 90
* Strength: 60 - 90
* Defence: 40 – 60
* Speed: 40 – 60
* Luck: 25% - 40%  

The project simulate a battle between **Orderus** and a wild beast, either at command line or using a web browser. On every battle, **Orderus** and the beast must be initialized with random properties, within their ranges.   
 
 The first attack is done by the player with the higher speed. If both players have the same speed, then the attack is carried on by the player with the highest luck. After an attack, the players switch roles: the attacker now defends, and the defender now attacks.  
   
The damage done by the attacker is calculated with the following formula:
```
Damage = Attacker strength – Defender defence
```
The damage is subtracted from the defender’s health. An attacker can miss their hit and do no damage if the defender gets lucky that turn.  

Orderus’ skills occur randomly, based on their chances, so take them into account on each turn.  

**Game over**  

The game ends when one of the players remain without health, or the number of turns reaches 20.  

The application must output the results each turn: what happened, which skills were used (if any), the damage done, defender’s health left.  

If we have a winner before the maximum number of rounds is reached, he must be declared.

## Install and run on local machine

### Prerequisites
Before start install, user will need:
* One [Bionic Beaver][1], and a non-root user with `sudo` privileges.
* [Composer][2]
* [Docker][3]
* [Docker Compose][4]

### Install local and run
* Clone repository
```shell script
git clone https://github.com/tudor-rusu/oop-game.git ${PROJECT_ROOT}
```
* Copy and adjust environmental settings in the root of the project, assumed `${PROJECT_ROOT}`:
```shell script
cp ${PROJECT_ROOT}/src/.env.dist ${PROJECT_ROOT}/src/.env
```
* Run the `sh` script which deploy Docker environment
```shell script
./run.sh
```

* Enjoy!

[conventional-commits-image]: https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg
[conventional-commits-url]: https://conventionalcommits.org/
[1]: http://releases.ubuntu.com/18.04.4/
[2]: https://www.digitalocean.com/community/tutorials/how-to-install-and-use-composer-on-ubuntu-18-04
[3]: https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-18-04
[4]: https://www.digitalocean.com/community/tutorials/how-to-install-docker-compose-on-ubuntu-18-04


CHANGELOG.md
## [0.2.1] - 2020-06-07

### Modified
- CHANGELOG.md
- Application component
- Controller abstraction
- ControllerFactory abstraction

## [0.2.0] - 2020-06-07

### Added
- AJAX implementation
- http error code views
- minimal layout
- DataManipulation trait

### Modified
- CHANGELOG.md
- views main layout
- project js and css
- main Model class
- Orderus and Beast models
- HomeController class
- PageController class
- Routing Trait

## [0.1.9.1] - 2020-06-05

### Added
- FlashMessages class

### Modified
- CHANGELOG.md
- views main layout
- project js
- main Cntroller class
- HomeController class - index method
- JsonConnection

## [0.1.9] - 2020-06-04

### Added
- interfaces
- data connection
- json data

### Modified
- CHANGELOG.md
- models structure
- Orderus model

## [0.1.8] - 2020-06-02

### Added
- Static Page Controller
- static page About view
- assets images

### Modified
- CHANGELOG.md
- project.css

## [0.1.7] - 2020-06-02

### Added
- Phpunit tests

### Modified
- CHANGELOG.md
- project.css
- run.sh

## [0.1.6] - 2020-06-01

### Added
- Routing Trait
- Main Menu Config

### Modified
- CHANGELOG.md
- abstractions Controller, ControllerFactory and Router
- HomeController

## [0.1.5] - 2020-06-01

### Added
- views
- assets

### Modified
- CHANGELOG.md

## [0.1.4] - 2020-06-01

### Added
- composer and phpunit support
- abstractions
- components
- controllers

### Modified
- CHANGELOG.md
- composer
- index.php

## [0.1.3] - 2020-05-31

### Modified
- CHANGELOG.md
- docker env files related to use different OS
- move .env inside src folder

## [0.1.2] - 2020-05-24

### Added
- php Dockerfile and ini
- separate config for php
- env
- example certificates

### Modified
- CHANGELOG.md
- docker-compose

### Removed
- docker conf become .env

## [0.1.1] - 2020-04-17

### Added
- SSL self-signed certification
- nginx Dockerfile
- separate config for nginx
- docker conf

### Modified
- CHANGELOG.md
- nginx script
- docker-compose

## [0.1.0] - 2020-04-17

### Added
- script files for project configuration
- script files for nginx container build
- script files for php container build
- script file with global vars and functions
- folders structure
- docker-compose.yml

### Modified
- gitignore, README.md and CHANGELOG.md

## [0.0.1] - 2020-04-17

### Added
- gitignore, README.md, LICENSE and CHANGELOG.md
- initial commit and structure

src/app/assets/js/project.js
window.addEventListener('load', function () {

    // auto remove alert after 7sec
    setTimeout(function () {
        let alertContainer = $('.alert');

        if (alertContainer.is(':visible')) {
            alertContainer.fadeOut();
        }
    }, 7000)

    // security token
    $.ajaxSetup({
        headers: {
            'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content')
        }
    });

    $('#fight').focus();

    // analise fight
    $('#fight').click(function (e) {
        e.preventDefault();
        let heroAttr              = $('#hero-attributes');
        let beastAttr             = $('#beast-attributes');
        let battleStats           = $('#battle-stats');
        let battleStatisticsTable = $('#battle-statistics');
        let turnSide              = $('#turn-side');
        let currentTurn           = $('#current-turn > small');
        let btnFightObj           = $('#fight');
        let initialText           = btnFightObj.html();
        let maxTurns              = battleStats.data('maxturns');
        let statisticsTurn        = 0;
        //set data
        let dataFight             = {
            turn: battleStats.data('turn'),
            hero: getPlayerData(heroAttr),
            beast: getPlayerData(beastAttr)
        };

        btnFightObj.html('<span class="spinner-border spinner-border-sm mr-2" role="status" aria-hidden="true"></span>' + initialText).addClass('disabled');

        $.ajax({
            type: "POST",
            url: '/home/analise',
            data: dataFight,
            success: function (data) {
                let fightResults = $.parseJSON(data);
                btnFightObj.html(initialText).removeClass('disabled');
                if ($.isEmptyObject(data.error)) {
                    // increase the turn
                    let newTurn = fightResults.turn + 1;
                    if (fightResults.winner) {
                        btnFightObj.hide()
                        $('#new-battle').show().focus();
                        $('#center-info').hide();
                        if (fightResults.winner === 'hero') {
                            $('#beast-dead').show();
                        } else {
                            $('#hero-dead').show();
                        }
                    } else {
                        currentTurn.html('turn<br />' + newTurn);
                        battleStats.data('turn', newTurn);
                        setTimeout(function () {
                            btnFightObj.click();
                        }, 500)
                    }

                    // show battle statistics
                    battleStatisticsTable.show();
                    let newRowContent = '<tr>' +
                        '<td>' + fightResults.turn + '</td>' +
                        '<td>' + (fightResults.stats.attacker) + '</td>' +
                        '<td>' + (fightResults.stats.defender_luck) + '</td>' +
                        '<td>' + (fightResults.stats.damage) + '</td>' +
                        '<td>' + (fightResults.stats.defender_health) + '</td>' +
                        '<td>' + (fightResults.stats.hero_skill) + '</td>' +
                        '<td>' + (fightResults.stats.winner) + '</td>' +
                        '</tr>';
                    battleStatisticsTable.find('tbody').append(newRowContent);

                    // set frontend values
                    let playersResultsArray = {
                        'hero': {
                            'domObj': heroAttr,
                            'action': fightResults.hero.action,
                            'luck': fightResults.hero.luck,
                            'health': fightResults.hero.health
                        },
                        'beast': {
                            'domObj': beastAttr,
                            'action': fightResults.beast.action,
                            'luck': fightResults.beast.luck,
                            'health': fightResults.beast.health
                        }
                    };

                    $('div[id$=-card]').removeClass(function (index, className) {
                        return (className.match(/(^|\s)border-\S+/g) || []).join(' ');
                    });

                    for (let key in playersResultsArray) {
                        setPlayerData(key, playersResultsArray[key]['domObj'], playersResultsArray[key]['action']);
                        showPlayerLuck(key, playersResultsArray[key]['luck']);
                        updatePlayerHealth(key, playersResultsArray[key]['health']);
                    }
                }
            }
        });
    });

    function getPlayerData(playerDomObj) {
        return {
            name: playerDomObj.data('player'),
            action: playerDomObj.data('action'),
            health: playerDomObj.find('tr[data-name="health"]').data('value'),
            strength: playerDomObj.find('tr[data-name="strength"]').data('value'),
            defence: playerDomObj.find('tr[data-name="defence"]').data('value'),
            speed: playerDomObj.find('tr[data-name="speed"]').data('value'),
            luck: playerDomObj.find('tr[data-name="luck"]').data('value')
        };
    }

    function setPlayerData(playerName, playerDomObj, playerReturnValue) {
        let playerCard        = $('#' + playerName + '-card');
        let playerFightStatus = $('#' + playerName + '-fight-status');
        let turnSide          = $('#turn-side');
        playerDomObj.data('action', playerReturnValue);

        switch (playerReturnValue) {
            case 0:
                playerCard.addClass('border-danger');
                playerFightStatus.html('defender');
                if (playerName === 'hero') {
                    turnSide.html('&#129092;');
                }
                break;
            case 1:
                playerCard.addClass('border-success');
                playerFightStatus.html('attacker');
                if (playerName === 'hero') {
                    turnSide.html('&#129094;');
                }
                break;
        }
    }

    function showPlayerLuck(playerName, playerReturnValue) {
        let playerDomObj = $('#' + playerName + '-luck');
        if (playerReturnValue === 1) {
            playerDomObj.show();
        } else {
            playerDomObj.hide();
        }
    }

    function updatePlayerHealth(playerName, playerReturnValue) {
        let playerHealth         = $('#' + playerName + '-attributes tbody tr[data-name="health"]');
        let playerHealthProgress = $('#' + playerName + '-attributes div.progress-bar');
        if (playerReturnValue >= 0) {
            playerHealth.data('value', playerReturnValue);
            playerHealth.find('.attr-value').html(playerReturnValue);
            playerHealthProgress.attr('style', 'width: ' + playerReturnValue + '%;')
                .attr('aria-valuenow', playerReturnValue);
            playerHealthProgress.removeClass(function (index, className) {
                return (className.match(/(^|\s)bg-\S+/g) || []).join(' ');
            });
            if (playerReturnValue > 75) {
                playerHealthProgress.addClass('bg-success');
            } else if (playerReturnValue > 50 && playerReturnValue < 74) {
                playerHealthProgress.addClass('bg-info');
            } else if (playerReturnValue > 25 && playerReturnValue < 49) {
                playerHealthProgress.addClass('bg-warning');
            } else if (playerReturnValue < 24) {
                playerHealthProgress.addClass('bg-danger');
            }
        }
    }

});